### PR TITLE
k2: allow serial port to be temporarily suspended

### PR DESCRIPTION
**Pull request by Risca**
I'm currently doing some development on a device (https://en.wikipedia.org/wiki/CHIP_(computer)) with a rather flaky wifi connection, so I want to be able to transfer files to and from a SUT using the serial port instead. This is possible if we temporarily suspend the log line handling of the serial port and let the _xmodem_ python module do the actual transfer.

### CODE REVIEW

File: k2/addons/k2-serial/zserial/serial.py [hunk 1]
```diff
@@ -54,6 +55,75 @@ def serial_exec(sut, messagebus):
...
+                    endmark='rz: ready to receive /tmp/foo')
+                with serial.suspended() as ser:
+                    def getc(size, timeout=1):
+                        return ser.read(size) or None
+
+                    def putc(data, timeout=1):
+                        return ser.write(data)
+
+                    modem = XMODEM1k(getc, putc)
+                    modem.send(f)
```
Comment (Risca) [inline k2/addons/k2-serial/zserial/serial.py:105]: This actually works =)

If the K2 interface addon for file transfer happen to re-appear, then I imagine XMODEM being just one of the ways to transfer files to a SUT.

File: k2/addons/k2-serial/zserial/serial.py [hunk 2]
```diff
@@ -118,7 +194,7 @@ def before_command(self, message, messagebus):
     @sequential_dispatcher([SERIAL_SEND_COMMAND], [SERIAL_ENDPOINT], entity_option_id=SUT)
     def send_serial_command(self, message):
         data = message.data
-        if isinstance(message.data, SendSerialCommandData):
+        if isinstance(data, SendSerialCommandData):
```
Comment (Risca) [inline k2/addons/k2-serial/zserial/serial.py:197]: This change is unrelated to this review and could be merged in any case.

File: k2/addons/k2-serial/zserial/serial.py (hunk 1 above)
Comment (per-bohlin) [reply k2/addons/k2-serial/zserial/serial.py:105]: I agree that we need a file-transfer mechanism that is done using a component-interface.
I will get back to you on this in a couple of days. I might be able to dig up the HTTP-server based image transfer from upstream.

File: k2/addons/k2-serial/zserial/serial.py [hunk 3]
```diff
@@ -54,6 +55,75 @@ def serial_exec(sut, messagebus):
...
+        self._serial_connection = serial_connection
+
+    @contextmanager
+    def suspended(self):
+        """Suspend the normal serial log line handling."""
+
+        try:
+            logger.debug(
+                'Suspending serial port for entity {entity}'.format(entity=self._sut.entity))
+            self._messagebus.trigger_event(SERIAL_SUSPEND, SERIAL_ENDPOINT, entity=self._sut.entity)
```
Comment (per-bohlin) [inline k2/addons/k2-serial/zserial/serial.py:120]: You need to use messagebus.send_request() and wait for the result to ensure that the zserial service has had time to process the event. Same goes for the resume in the finally-block. Example of using mesagebus-requests can be found in connectioncheck.py:89.

**Comment by per-bohlin** [comment]
I didn't have time to look at everything yet. I will continue next week.

File: k2/addons/k2-serial/zserial/serial.py (hunk 1 above)
Comment (Risca) [reply k2/addons/k2-serial/zserial/serial.py:105]: That would be awesome! :smiley:

File: k2/addons/k2-serial/zserial/serial.py (hunk 3 above)
Comment (Risca) [reply k2/addons/k2-serial/zserial/serial.py:120]: I found (and understood) the `MessageBus::send_request()` only after I pushed this up. I'm thinking of having the request return a raw serial port as the result of the request. Then I don't need the `RawSerialPort` component at all.

**Comment by Risca** [comment]
> You need to use messagebus.send_request() and wait for the result to ensure that the zserial service has had time to process the event. Same goes for the resume in the finally-block.

Fixed in latest version :smiley: I also could get rid of the "helper component" for getting the Serial instance.

File: k2/addons/k2-serial/zserial/connection.py [hunk 4]
```diff
@@ -27,29 +27,48 @@ class SerialConnectionError(Exception):
...
         self.thread = thread
         self.connection = connection
+        self.serial = serial
+        self.protocol_factory = protocol_factory
 
     def close(self):
         self.thread.close()
 
+    def suspend(self):
+        self.thread.stop()
```
Comment (per-bohlin) [inline k2/addons/k2-serial/zserial/connection.py:40]: Why use .stop() instead of .close()?
Close seems to do more things and at first glans looks more appropriate since we overwriting the thread-instance when we resume.

Comment (per-bohlin) [reply k2/addons/k2-serial/zserial/connection.py:40]: Never mind, reading the documentation of ReaderThread a bit more careful, it seems the only way is to use .stop() and create a new instance. So ignore above comment.

File: k2/addons/k2-serial/zserial/connection.py [hunk 5]
```diff
@@ -27,29 +27,48 @@ class SerialConnectionError(Exception):
...
+        if self.thread.is_alive():
+            raise SerialConnectionError('Failed to suspend serial thread')
+
+    def resume(self):
+        self.thread = ReaderThread(self.serial, self.protocol_factory)
+        self.connection = self.thread.__enter__()
+
+    def is_suspended(self):
+        return False if self.thread.is_alive() else True
```
Comment (per-bohlin) [inline k2/addons/k2-serial/zserial/connection.py:49]: `return not self.thread.is_alive()`
?

File: k2/addons/k2-serial/zserial/serial.py [hunk 6]
```diff
@@ -54,6 +55,75 @@ def serial_exec(sut, messagebus):
...
+        try:
+            logger.debug(
+                'Suspending serial port for entity {entity}'.format(entity=self._sut.entity))
+            port = self._messagebus.send_request(
+                SERIAL_SUSPEND, SERIAL_ENDPOINT, entity=self._sut.entity).wait()[0].result()
+            yield port
+        finally:
+            logger.debug('Resuming serial port for entity {entity}'.format(entity=self._sut.entity))
+            self._messagebus.send_request(
+                SERIAL_RESUME, SERIAL_ENDPOINT, entity=self._sut.entity).wait()[0].result()
```
Comment (per-bohlin) [inline k2/addons/k2-serial/zserial/serial.py:124]: The implementation can be extracted into two methods `.suspend()` and `.resume()`, called from the try- and finally-block.
That way the component can be used without using the context manager and without resorting to ugliness such as `.suspended().__enter__()`.

File: k2/addons/k2-serial/zserial/serial.py (hunk 2 above)
Comment (per-bohlin) [reply k2/addons/k2-serial/zserial/serial.py:192]: good

File: k2/addons/k2-serial/zserial/connection.py [hunk 7]
```diff
@@ -27,29 +27,48 @@ class SerialConnectionError(Exception):
...
     try:
         serial = Serial(port, baudrate, rtscts=virtual, dsrdtr=virtual, timeout=timeout)
         _lock_serial_port(serial)
-        serial_thread = ReaderThread(serial, _serial_connection(messagebus, entity, filters))
+        protocol_factory = _serial_connection(messagebus, entity, filters)
+        serial_thread = ReaderThread(serial, protocol_factory)
 
-        return SerialConnectionWrapper(serial_thread, serial_thread.__enter__())
+        return SerialConnectionWrapper(
+            serial_thread, serial_thread.__enter__(), serial, protocol_factory)
```
Comment (per-bohlin) [inline k2/addons/k2-serial/zserial/connection.py:71]: Create a ConnectionReaderThreadFactory of a sort that handles the creation and returns the connection and readerthread. It can be done by an internal function or closure that has access to all variables needed.
That way the SerialConnectionWrapper doesn't need to know how to recreate the thread and instead just call the factory. That eliminates the need for all four arguments and replaces it with one - the factory.

File: k2/addons/k2-serial/zserial/serial.py [hunk 8]
```diff
@@ -54,6 +55,75 @@ def serial_exec(sut, messagebus):
...
+@requires(sut='Sut', can=['serial'])
+@requires(messagebus='MessageBus')
+@component(name='RawSerialPort', can=['serial'], provided_by_extension='zserial')
+class RawSerialPort(object):
+    """
+    Access the serial port in raw byte mode.
+
+    Temporarily suspend the packetized log line handling to provide raw byte
+    access to the serial port.
+
```
Comment (per-bohlin) [inline k2/addons/k2-serial/zserial/serial.py:67]: Add a warning that this mechanism is not thread-safe - in multiple aspects.
Suspension and resuming can be done by anyone. Which means that if multiple entities has used this feature overlapping in time, the first resume will destroy for the others.
Also, since multiple entities can suspend, the port is not guaranteed to be free. It can be used by other "suspenders".

Consider if the suspend-dispatcher should be blocking, and not return if it has been suspended by another context, until the service is resumed. Alternatively, that it raises an exception if you try to suspend if already suspended.
The minimum would at least to warn about it in the documentation. If you can find a nice solution for it is a plus.

**Review by per-bohlin — COMMENTED** [review]
A couple of things that need addressing. A good feature to have but not unproblematic to implement.

File: k2/addons/k2-serial/zserial/serial.py (hunk 2 above)
Comment (Risca) [reply k2/addons/k2-serial/zserial/serial.py:192]: Added to master in cac4b28e9d576144faa43e77646bab1a4090dcd7

File: k2/addons/k2-serial/zserial/connection.py (hunk 4 above)
Comment (Risca) [reply k2/addons/k2-serial/zserial/connection.py:40]: Yes, I wanted to keep the serial connection open so as to not loose any potential data, and "restarting" a thread in python is a no-go.

File: k2/addons/k2-serial/zserial/connection.py (hunk 5 above)
Comment (Risca) [reply k2/addons/k2-serial/zserial/connection.py:49]: Can't remember why I did it like that, but your suggestion seem to work now so I'll go with that.

File: k2/addons/k2-serial/zserial/connection.py (hunk 7 above)
Comment (Risca) [reply k2/addons/k2-serial/zserial/connection.py:71]: Agreed. Will fix in the coming days.

File: k2/addons/k2-serial/zserial/serial.py (hunk 8 above)
Comment (Risca) [reply k2/addons/k2-serial/zserial/serial.py:67]: Agreed. I'll see what I can do about the thread safety

File: k2/addons/k2-serial/zserial/serial.py (hunk 6 above)
Comment (Risca) [reply k2/addons/k2-serial/zserial/serial.py]: Good suggestion! Fixed in latest diff :wink:

File: k2/addons/k2-serial/zserial/connection.py (hunk 7 above)
Comment (Risca) [reply k2/addons/k2-serial/zserial/connection.py]: Fixed now. I'll take a look at the other stuff shortly :smiley:

**Comment by Risca** [comment]
This change seem to break unit tests, so I'll have to look into that as well

File: k2/addons/k2-serial/zserial/serial.py (hunk 8 above)
Comment (Risca) [reply k2/addons/k2-serial/zserial/serial.py:67]: I added exceptions for now. I don't know if blocking is the way to go. E.g. transferring a large file @ 9600 baud would block someone else for a really long time. Maybe that's ok, but maybe that will trigger a timeout exception in some other (future) component waiting to suspend, in which case it would have been easier to troubleshoot if the SerialExtension had thrown an exception right away? :shrug:
These are all hypothetical scenarios. I suggest we leave it at throwing exceptions when trying to suspend when already suspended and when trying to resume when not suspended.

**Comment by Risca** [comment]
I think this pull request is ready for review now. I might add some more tests for the suspend functionality, but I might as well do that as a separate pull request.

File: k2/addons/k2-serial/zserial/serial.py (hunk 1 above)
Comment (Risca) [reply k2/addons/k2-serial/zserial/serial.py:106]: @per-bohlin, any update on the HTTP-server based image transfer from upstream?

File: k2/addons/k2-serial/zserial/serial.py (hunk 8 above)
Comment (Risca) [reply k2/addons/k2-serial/zserial/serial.py:67]: I consider throwing an exception good enough, so I'm resolving this conversation :slightly_smiling_face:

**Comment by Risca** [comment]
This pull request is ready for review and possible merge.

File: k2/addons/k2-serial/zserial/serial.py (hunk 1 above)
Comment (per-bohlin) [reply k2/addons/k2-serial/zserial/serial.py:106]: Sorry, I've been very busy the last couple of weeks.
I'll prioritize to get the reviews done, then I'll look into finding the HTTP-server.

**Review by per-bohlin — APPROVED** [review]
Go ahead and merge.
Thanks!
